### PR TITLE
ENH: seed kwd for basinhopping. Closes #6278

### LINKED ISSUE #6278: optimize.basinhopping should take a RandomState object
... to ensure reproducibility of results.

### PR DESCRIPTION
**Pull request by andyfaff**
Adds `seed` kwd to `optimize.basinhopping` for repeatable minimisation.

### CODE REVIEW

**Comment by andyfaff** [comment]
Note: the `test__differential_evolution.TestDifferentialEvolutionSolver.test_best_solution_retrieval` test started failing when the changes were made to basinhopping. I'm bemused by this because it's not related code. Failure was probably due to a slightly different solution (caused by slightly different random number generator state) which fails the `assert_equal` test when the best solution is scaled. I relaxed to `assert_almost_equal` in the test.

File: scipy/optimize/_basinhopping.py [hunk 1]
```diff
@@ -248,12 +249,22 @@ class RandomDisplacement(object):
...
+
+    Parameters
+    ----------
+    stepsize : float, optional
+        stepsize
+    rand_state : None or `np.random.RandomState` instance, optional
+        The random number generator that generates the displacements
     """
-    def __init__(self, stepsize=0.5):
+    def __init__(self, stepsize=0.5, rand_state=None):
```
Comment (ev-br) [inline scipy/optimize/_basinhopping.py]: Bikeshedding: scikit-learn names this argument `random_state`. scipy.stats distributions follow. Maybe it's better to be consistent here.

File: scipy/optimize/_basinhopping.py [hunk 2]
```diff
@@ -248,12 +249,22 @@ class RandomDisplacement(object):
...
         self.stepsize = stepsize
+        self.rand_state = rand_state
+        if rand_state is None:
+            self.rand_state = np.random.mtrand._rand
```
Comment (ev-br) [inline scipy/optimize/_basinhopping.py]: Maybe use the utility `check_random_state`, https://github.com/scipy/scipy/blob/master/scipy/_lib/_util.py#L162
(which is also borrowed from scikit-learn)

**Comment by ev-br** [comment]
+1 for the feature.

File: scipy/optimize/_basinhopping.py (hunk 2 above)
Comment (andyfaff) [inline scipy/optimize/_basinhopping.py]: I didn't know about `check_random_state`, thank you. I've made the suggested changes.

File: scipy/optimize/_basinhopping.py [hunk 3]
```diff
@@ -248,12 +249,22 @@ class RandomDisplacement(object):
...
+        stepsize
+    random_state : None or `np.random.RandomState` instance, optional
+        The random number generator that generates the displacements
     """
-    def __init__(self, stepsize=0.5):
+    def __init__(self, stepsize=0.5, random_state=None):
         self.stepsize = stepsize
+        self.random_state = random_state
+        if random_state is None:
+            self.random_state = np.random.mtrand._rand
```
Comment (andyfaff) [inline scipy/optimize/_basinhopping.py]: Good catch, this is why code review is important. Changes made.

**Comment by ev-br** [comment]
This can probably be advertised in the release notes, but it can be done separately as well. Thanks Andrew, merging.

**Pull request merged**